<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Rishi Das Roy">
<meta name="dcterms.date" content="2025-01-09">
<meta name="description" content="A short evaluation of doublet detection methods">

<title>Rishi Das Roy - scATAC-seq: Doublet detection:</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-5ELK5DXSS2"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-5ELK5DXSS2', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"headline",
  "consent_type":"implied",
  "palette":"dark",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Rishi Das Roy</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../listing.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools tools-wide">
    <a href="https://www.linkedin.com/in/rishi-das-roy-80806a7/" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-linkedin"></i></a>
    <a href="https://x.com/aarsho" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#motivation" id="toc-motivation" class="nav-link active" data-scroll-target="#motivation"><span class="header-section-number">1</span> Motivation</a></li>
  <li><a href="#workflow" id="toc-workflow" class="nav-link" data-scroll-target="#workflow"><span class="header-section-number">2</span> Workflow</a>
  <ul class="collapse">
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="header-section-number">2.1</span> Data</a></li>
  <li><a href="#required-r-packages" id="toc-required-r-packages" class="nav-link" data-scroll-target="#required-r-packages"><span class="header-section-number">2.2</span> Required R packages</a></li>
  <li><a href="#read-data-as-seurat-object" id="toc-read-data-as-seurat-object" class="nav-link" data-scroll-target="#read-data-as-seurat-object"><span class="header-section-number">2.3</span> Read data as Seurat Object</a></li>
  <li><a href="#qc-metrics" id="toc-qc-metrics" class="nav-link" data-scroll-target="#qc-metrics"><span class="header-section-number">2.4</span> QC Metrics</a>
  <ul class="collapse">
  <li><a href="#computing" id="toc-computing" class="nav-link" data-scroll-target="#computing"><span class="header-section-number">2.4.1</span> Computing</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization"><span class="header-section-number">2.4.2</span> Visualization</a></li>
  <li><a href="#sec-Doublet-Identification" id="toc-sec-Doublet-Identification" class="nav-link" data-scroll-target="#sec-Doublet-Identification"><span class="header-section-number">2.4.3</span> Doublet Identification</a></li>
  <li><a href="#compare-scdblfinder-and-amulet" id="toc-compare-scdblfinder-and-amulet" class="nav-link" data-scroll-target="#compare-scdblfinder-and-amulet"><span class="header-section-number">2.4.4</span> Compare scDblFinder and amulet</a></li>
  </ul></li>
  <li><a href="#removing-bad-cells" id="toc-removing-bad-cells" class="nav-link" data-scroll-target="#removing-bad-cells"><span class="header-section-number">2.5</span> Removing bad cells</a></li>
  <li><a href="#normalization-and-linear-dimensional-reduction" id="toc-normalization-and-linear-dimensional-reduction" class="nav-link" data-scroll-target="#normalization-and-linear-dimensional-reduction"><span class="header-section-number">2.6</span> Normalization and linear dimensional reduction</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span class="header-section-number">2.7</span> Clustering</a></li>
  <li><a href="#sec-results" id="toc-sec-results" class="nav-link" data-scroll-target="#sec-results"><span class="header-section-number">2.8</span> Result</a>
  <ul class="collapse">
  <li><a href="#doublet-cells-in-the-clusters" id="toc-doublet-cells-in-the-clusters" class="nav-link" data-scroll-target="#doublet-cells-in-the-clusters"><span class="header-section-number">2.8.1</span> Doublet cells in the clusters??</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">2.8.2</span> References</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">scATAC-seq: Doublet detection:</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Single cell</div>
    <div class="quarto-category">ATAC-seq</div>
    <div class="quarto-category">Biology</div>
    <div class="quarto-category">analysis</div>
  </div>
  </div>

<div>
  <div class="description">
    A short evaluation of doublet detection methods
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Rishi Das Roy </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="motivation" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Motivation</h1>
<p>Here’s a revised version tailored for motivating the evaluation of scATAC-seq doublet detection methods:</p>
<hr>
<p>Chapter 23 of <a href="https://www.sc-best-practices.org/chromatin_accessibility/introduction.html">sc-best-practices.org</a> <span class="citation" data-cites="heumos2023best">(<a href="#ref-heumos2023best" role="doc-biblioref">Heumos et al. 2023</a>)</span> highlights and compares several methods for analyzing scATAC-seq data, identifying <code>Signac</code> <span class="citation" data-cites="Signac2021">(<a href="#ref-Signac2021" role="doc-biblioref">Stuart et al. 2021</a>)</span> as one of the leading tools for this purpose. However, <code>Signac</code> does not natively include a dedicated method for doublet detection. Instead, it recommends using the “total number of fragments in peaks” as a heuristic approach to flag potential doublets.</p>
<p>In contrast, the R package <code>scDblFinder</code> <span class="citation" data-cites="scDblFinder2022">(<a href="#ref-scDblFinder2022" role="doc-biblioref">Germain et al. 2022</a>)</span> provides a more systematic method for doublet detection, but its input requirements differ from the data format used by <code>Signac</code>. While sc-best-practices.org demonstrates combining these tools with additional Python packages, this approach introduces complexity to the workflow, making it less accessible for R-focused pipelines.</p>
<p>To address this gap, it is worth investigating the impact of doublets on classical PBMC datasets analyzed with <code>Signac</code>, while staying entirely within the R environment. Such an evaluation would not only streamline the workflow but also provide insights into the effectiveness of these tools in detecting and mitigating doublets in scATAC-seq data.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Do we need python for scATAC-seq?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The sc-best-practices.org demonstrates scATAC-seq analysis using both python and R packages. The use of python packages for reading the data and visualization is unnecessarily increase complexity of the workflow. In my opinion, the scATAC-seq analysis should be done solely through R packages.</p>
</div>
</div>
</div>
</section>
<section id="workflow" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Workflow</h1>
<p>This is based primarily on <a href="https://stuartlab.org/signac/articles/pbmc_vignette">Signac :: Analyzing PBMC scATAC-seq</a> vignette where <code>scDblFinder</code> methods are fused after computing QC metrics. If you are only interested into the doublet identification then jump to <a href="#sec-Doublet-Identification" class="quarto-xref">Section&nbsp;2.4.3</a> …</p>
<section id="data" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="data"><span class="header-section-number">2.1</span> Data</h2>
<p>The data used here from 10x Genomics same as in the original vignette.</p>
<details>
<summary>
Download command
</summary>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># -nc if file does not exist</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wget</span> <span class="at">-nc</span> https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wget</span> <span class="at">-nc</span> https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wget</span> <span class="at">-nc</span> https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wget</span> <span class="at">-nc</span> https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz.tbi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>File ‘10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5’ already there; not retrieving.

File ‘10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv’ already there; not retrieving.

File ‘10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz’ already there; not retrieving.

File ‘10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz.tbi’ already there; not retrieving.</code></pre>
</div>
</div>
</details>
</section>
<section id="required-r-packages" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="required-r-packages"><span class="header-section-number">2.2</span> Required R packages</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"Signac"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setRepositories</span>(<span class="at">ind=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="co"># needed to automatically install Bioconductor dependencies</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"Signac"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scDblFinder)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggvenn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="read-data-as-seurat-object" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="read-data-as-seurat-object"><span class="header-section-number">2.3</span> Read data as Seurat Object</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">Read10X_h5</span>(<span class="at">filename =</span> <span class="st">"10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">"10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv"</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="dv">1</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>chrom_assay <span class="ot">&lt;-</span> <span class="fu">CreateChromatinAssay</span>(</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> counts,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">sep =</span> <span class="fu">c</span>(<span class="st">":"</span>, <span class="st">"-"</span>),</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">fragments =</span> <span class="st">"10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz"</span>,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.cells =</span> <span class="dv">10</span>,</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.features =</span> <span class="dv">200</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> chrom_assay,</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"peaks"</span>,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">meta.data =</span> metadata,</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"PBMC"</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the features that correspond to standard chromosomes. </span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>peaks.keep <span class="ot">&lt;-</span> <span class="fu">seqnames</span>(<span class="fu">granges</span>(pbmc)) <span class="sc">%in%</span> <span class="fu">standardChromosomes</span>(<span class="fu">granges</span>(pbmc))</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> pbmc[<span class="fu">as.vector</span>(peaks.keep), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="qc-metrics" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="qc-metrics"><span class="header-section-number">2.4</span> QC Metrics</h2>
<section id="computing" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="computing"><span class="header-section-number">2.4.1</span> Computing</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute nucleosome signal score per cell</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">NucleosomeSignal</span>(<span class="at">object =</span> pbmc)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># compute TSS enrichment score per cell</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">TSSEnrichment</span>(<span class="at">object =</span> pbmc)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add fraction of reads in peaks</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>pct_reads_in_peaks <span class="ot">&lt;-</span> pbmc<span class="sc">$</span>peak_region_fragments <span class="sc">/</span> pbmc<span class="sc">$</span>passed_filters <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># add blacklist ratio</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>blacklist_ratio <span class="ot">&lt;-</span> <span class="fu">FractionCountsInRegion</span>(</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc, </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">'peaks'</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">regions =</span> blacklist_hg38_unified</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="visualization" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="visualization"><span class="header-section-number">2.4.2</span> Visualization</h3>
<details>
<summary>
Basic QC metrics
</summary>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DensityScatter</span>(pbmc, <span class="at">x =</span> <span class="st">'nCount_peaks'</span>, <span class="at">y =</span> <span class="st">'TSS.enrichment'</span>, <span class="at">log_x =</span> <span class="cn">TRUE</span>, <span class="at">quantiles =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-quantiles" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-quantiles-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="scATAC_seq_files/figure-html/fig-quantiles-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" data-glightbox="description: .lightbox-desc-1"><img src="scATAC_seq_files/figure-html/fig-quantiles-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-quantiles-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Relationship between Peaks and TSS enrichment.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>nucleosome_group <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(pbmc<span class="sc">$</span>nucleosome_signal <span class="sc">&gt;</span> <span class="dv">4</span>, <span class="st">'NS &gt; 4'</span>, <span class="st">'NS &lt; 4'</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FragmentHistogram</span>(<span class="at">object =</span> pbmc, <span class="at">group.by =</span> <span class="st">'nucleosome_group'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">c</span>(<span class="st">'nCount_peaks'</span>, <span class="st">'TSS.enrichment'</span>, <span class="st">'blacklist_ratio'</span>, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">'nucleosome_signal'</span>, <span class="st">'pct_reads_in_peaks'</span>),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt.size =</span> <span class="fl">0.1</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">5</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-vlnplot" class="quarto-figure quarto-figure-center quarto-float anchored" width="1152">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-vlnplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="scATAC_seq_files/figure-html/fig-vlnplot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" data-glightbox="description: .lightbox-desc-2"><img src="scATAC_seq_files/figure-html/fig-vlnplot-1.png" id="fig-vlnplot" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-vlnplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2
</figcaption>
</figure>
</div>
</div>
</div>
</details>
</section>
<section id="sec-Doublet-Identification" class="level3" data-number="2.4.3">
<h3 data-number="2.4.3" class="anchored" data-anchor-id="sec-Doublet-Identification"><span class="header-section-number">2.4.3</span> Doublet Identification</h3>
<p>The <code>scDblFinder</code> package <span class="citation" data-cites="scDblFinder2022">(<a href="#ref-scDblFinder2022" role="doc-biblioref">Germain et al. 2022</a>)</span> offers two distinct methods for doublet detection: <code>scDblFinder</code> and <code>amulet</code> <span class="citation" data-cites="thibodeau2021amulet">(<a href="#ref-thibodeau2021amulet" role="doc-biblioref">Thibodeau et al. 2021</a>)</span>. However, these methods require different input formats—<code>scDblFinder</code> uses a <code>SingleCellExperiment</code> object <span class="citation" data-cites="SingleCellExperiment2020">(<a href="#ref-SingleCellExperiment2020" role="doc-biblioref">Amezquita et al. 2020</a>)</span>, while <code>amulet</code> operates on raw data files. In this analysis, the results from both methods are seamlessly integrated back into the original <code>Seurat</code> data object, allowing for continued analysis within a unified <code>Seurat</code> framework.</p>
<section id="with-scdblfinder" class="level4" data-number="2.4.3.1">
<h4 data-number="2.4.3.1" class="anchored" data-anchor-id="with-scdblfinder"><span class="header-section-number">2.4.3.1</span> with scDblFinder</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">scDblFinder</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.SingleCellExperiment</span>(pbmc),</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">artificialDoublets =</span> <span class="dv">1</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">aggregateFeatures =</span> <span class="cn">TRUE</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">nfeatures =</span> <span class="dv">25</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">processing =</span> <span class="st">"normFeatures"</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="with-amulet" class="level4" data-number="2.4.3.2">
<h4 data-number="2.4.3.2" class="anchored" data-anchor-id="with-amulet"><span class="header-section-number">2.4.3.2</span> with Amulet</h4>
<p>“As in the original implementation, we recommend excluding the mitochondrial and sex chromosomes, as well as repetitive regions. This can be specified with the <code>?regionsToExclude</code> argument (see the underlying <code>?getFragmentOverlaps</code>). It can be used as follows:” ~ <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/scDblFinder/inst/doc/scATAC.html#using-the-amulet-method">scDblFinder Vignet</a></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The ‘amulet’ function is very slow and has very large memory foot print. Use arguments fullInMemory and BPPARAM in the presence of large memory and multiple cpus.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>fragfile <span class="ot">&lt;-</span> <span class="st">"10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz"</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>repeats <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr6"</span>, <span class="fu">IRanges</span>(<span class="dv">1000</span>,<span class="dv">2000</span>))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># it's better to combine these with mitochondrial and sex chromosomes</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>otherChroms <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="fu">c</span>(<span class="st">"M"</span>,<span class="st">"chrM"</span>,<span class="st">"MT"</span>,<span class="st">"X"</span>,<span class="st">"Y"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>),<span class="fu">IRanges</span>(<span class="dv">1</span>L,<span class="at">width=</span><span class="dv">10</span><span class="sc">^</span><span class="dv">8</span>))</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># here since I don't know what chromosome notation you'll be using I've just put them all,</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># although this will trigger a warning when combining them:</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>toExclude <span class="ot">&lt;-</span> <span class="fu">suppressWarnings</span>(<span class="fu">c</span>(repeats, otherChroms))</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># we then launch the method</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Use fullInMemory=TRUE iff large memory available</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>res_amulet <span class="ot">&lt;-</span> <span class="fu">amulet</span>(fragfile, <span class="at">regionsToExclude=</span>toExclude, </span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fullInMemory=</span><span class="cn">TRUE</span>, </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                     <span class="at">BPPARAM=</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results are as a data frame where the <em>q-values</em> column is useful to filter out doublet cells.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>                   nFrags uniqFrags nAbove2 total.nAbove2    p.value   q.value
AAACGAAAGAGAGGTA-1  22250     22250      14            59 0.51612176 0.9999996
AAACGAAAGCAGGAGG-1   8879      8879       0             5 0.99999963 0.9999996
AAACGAAAGGAAGAAC-1  27740     27740      17            50 0.23633091 0.9848211
AAACGAAAGTCGACCC-1  28689     28689      23            81 0.01722372 0.1135366
AAACGAACAAGCACTT-1  12973     12973       3            24 0.99975592 0.9999996
AAACGAACAAGCGGTA-1  33812     33812      14            50 0.51612176 0.9999996</code></pre>
</div>
</div>
</section>
<section id="integrate-doublet-information" class="level4" data-number="2.4.3.3">
<h4 data-number="2.4.3.3" class="anchored" data-anchor-id="integrate-doublet-information"><span class="header-section-number">2.4.3.3</span> Integrate doublet information</h4>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Evaluate Amulet with scDblFinder</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>meta_scdblfinder <span class="ot">&lt;-</span> sce<span class="sc">@</span>colData<span class="sc">@</span>listData <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">'scDblFinder'</span>)) <span class="co"># 'scDblFinder.class')</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(meta_scdblfinder) <span class="ot">&lt;-</span> sce<span class="sc">@</span>colData<span class="sc">@</span>rownames</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>renamed_amulet <span class="ot">&lt;-</span> res_amulet <span class="sc">%&gt;%</span> <span class="fu">setNames</span>(<span class="fu">paste0</span>(<span class="st">'amulet.'</span>, <span class="fu">names</span>(.)))</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>combo_scdblfinder_amulet <span class="ot">&lt;-</span> <span class="fu">merge</span>(meta_scdblfinder, renamed_amulet, <span class="at">by=</span><span class="dv">0</span>, <span class="at">all.x=</span><span class="cn">TRUE</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(combo_scdblfinder_amulet) <span class="ot">&lt;-</span> combo_scdblfinder_amulet<span class="sc">$</span>Row.names</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore results and add to seurat object</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> combo_scdblfinder_amulet <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">'scDblFinder.class'</span>, <span class="st">'amulet.q.value'</span>)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>amulet_class <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fifelse</span>(pbmc<span class="sc">$</span>amulet.q.value <span class="sc">&gt;</span> <span class="fl">0.05</span>, <span class="st">'singlet'</span>, <span class="st">'doublet'</span>,<span class="at">na=</span><span class="st">"singlet"</span>)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>amulet_class <span class="ot">&lt;-</span> <span class="fu">factor</span>(pbmc<span class="sc">$</span>amulet_class, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">'singlet'</span>, <span class="st">'doublet'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="compare-scdblfinder-and-amulet" class="level3" data-number="2.4.4">
<h3 data-number="2.4.4" class="anchored" data-anchor-id="compare-scdblfinder-and-amulet"><span class="header-section-number">2.4.4</span> Compare scDblFinder and amulet</h3>
<div class="cell preview-image">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>p_scdbl_amu <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(<span class="fu">list</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">scDblFinder =</span> <span class="fu">rownames</span>(meta_scdblfinder[meta_scdblfinder<span class="sc">$</span>scDblFinder.class <span class="sc">==</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">"doublet"</span>, ]),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">amulet =</span> <span class="fu">rownames</span>(res_amulet[res_amulet<span class="sc">$</span>q.value <span class="sc">&lt;</span> <span class="fl">0.05</span>, ]),</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">TotalCells =</span> <span class="fu">colnames</span>(pbmc)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>),  <span class="at">text_size =</span> <span class="dv">3</span>, <span class="at">set_name_size =</span> <span class="dv">3</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>p_amu_cor <span class="ot">&lt;-</span> pbmc<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggscatter</span>(<span class="at">x=</span><span class="st">"nCount_peaks"</span>, <span class="at">y =</span> <span class="st">"amulet.q.value"</span>, </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="fl">0.1</span>, <span class="at">rug =</span> <span class="cn">TRUE</span>)<span class="sc">+</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">yscale</span>(<span class="st">"log10"</span>, <span class="at">.format =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Total number of fragments in peak"</span>, <span class="at">y=</span><span class="fu">expression</span>(<span class="st">"log"</span>[<span class="dv">10</span>]<span class="sc">*</span><span class="st">"q-value"</span>), <span class="at">title=</span><span class="st">"amulet"</span>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>p_scDbl <span class="ot">&lt;-</span> pbmc<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggviolin</span>( <span class="at">y=</span><span class="st">"nCount_peaks"</span>,  <span class="at">fill=</span><span class="st">"scDblFinder.class"</span>) <span class="sc">+</span><span class="fu">xlab</span>(<span class="st">"scDblFinder"</span>)<span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ylab</span>(<span class="st">"Total number of</span><span class="sc">\n</span><span class="st">fragments in peak"</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>p_amu <span class="ot">&lt;-</span> pbmc<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggviolin</span>( <span class="at">y=</span><span class="st">"nCount_peaks"</span>,  <span class="at">fill=</span><span class="st">"amulet_class"</span>)<span class="sc">+</span><span class="fu">xlab</span>(<span class="st">"amulet"</span>)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(p_scdbl_amu, </span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>          <span class="fu">ggarrange</span>(p_amu_cor, </span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">ggarrange</span>(p_scDbl<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"y.text"</span>)<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"legend.title"</span>)<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"x.text"</span>), </span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>                              p_amu<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"y.text"</span>)<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"y.title"</span>)<span class="sc">+</span><span class="fu">rremove</span>(<span class="st">"x.text"</span>),</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>                              <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend =</span> <span class="st">"bottom"</span>),</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>          )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-doublet_amulet" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-doublet_amulet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="scATAC_seq_files/figure-html/fig-doublet_amulet-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" data-glightbox="description: .lightbox-desc-3"><img src="scATAC_seq_files/figure-html/fig-doublet_amulet-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-doublet_amulet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Amulet more correlats with ‘Total number of fragments in peaks’
</figcaption>
</figure>
</div>
</div>
</div>
<p>The amulet method identifies more doublets than scDblFinder. Around 13% cells are doublet and amulet score correlates with “Total number of fragments in peak”. The cells &lt; 0.05 amulet q-values are marked as doublet which shows a sharp cut off compare scDblFinder. Overall, it looks amulet has better performance than the other one.</p>
</section>
</section>
<section id="removing-bad-cells" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="removing-bad-cells"><span class="header-section-number">2.5</span> Removing bad cells</h2>
<p>Using the same threshold as used in Signac.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> pbmc,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">subset =</span> nCount_peaks <span class="sc">&gt;</span> <span class="dv">9000</span> <span class="sc">&amp;</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    nCount_peaks <span class="sc">&lt;</span> <span class="dv">100000</span> <span class="sc">&amp;</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    pct_reads_in_peaks <span class="sc">&gt;</span> <span class="dv">40</span> <span class="sc">&amp;</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    blacklist_ratio <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">&amp;</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    nucleosome_signal <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">&amp;</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    TSS.enrichment <span class="sc">&gt;</span> <span class="dv">4</span> </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># amulet_group == "Singlet"</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>pbmc_signac</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="normalization-and-linear-dimensional-reduction" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="normalization-and-linear-dimensional-reduction"><span class="header-section-number">2.6</span> Normalization and linear dimensional reduction</h2>
<!--#  The combined steps of TF-IDF followed by SVD are known as latent
semantic indexing (LSI)-->
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">RunTFIDF</span>(pbmc_signac)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">FindTopFeatures</span>(pbmc_signac, <span class="at">min.cutoff =</span> <span class="st">'q0'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">RunSVD</span>(pbmc_signac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="clustering" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="clustering"><span class="header-section-number">2.7</span> Clustering</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The first dimension will be removed from the downstream analysis due to high correlation with sequencing depth.</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(<span class="at">object =</span> pbmc_signac, <span class="at">reduction =</span> <span class="st">'lsi'</span>, <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(<span class="at">object =</span> pbmc_signac, <span class="at">reduction =</span> <span class="st">'lsi'</span>, <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>pbmc_signac <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(<span class="at">object =</span> pbmc_signac, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">algorithm =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-results" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="sec-results"><span class="header-section-number">2.8</span> Result</h2>
<section id="doublet-cells-in-the-clusters" class="level3" data-number="2.8.1">
<h3 data-number="2.8.1" class="anchored" data-anchor-id="doublet-cells-in-the-clusters"><span class="header-section-number">2.8.1</span> Doublet cells in the clusters??</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>p_signac_filter <span class="ot">&lt;-</span> pbmc<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">y=</span>nCount_peaks, <span class="at">x=</span>TSS.enrichment, <span class="at">color=</span>amulet_group))<span class="sc">+</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">c</span>(<span class="dv">9000</span>,<span class="dv">100000</span>),<span class="at">color=</span><span class="st">"darkblue"</span>)<span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">y=</span><span class="dv">92000</span>, <span class="at">x=</span><span class="dv">10</span>, <span class="at">label=</span><span class="st">"Signac cut-off"</span>, <span class="at">angle=</span><span class="dv">0</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">color=</span><span class="st">"darkblue"</span>)<span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.1</span>)<span class="sc">+</span><span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">"amulet"</span>))<span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()<span class="sc">+</span><span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"top"</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>p_cluster <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(<span class="at">object =</span> pbmc_signac, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>p_cluster_amulet <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(<span class="at">object =</span> pbmc_signac,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>                            <span class="at">group.by =</span>  <span class="st">"amulet_class"</span>,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>                            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>p_cluster_scDblFinder <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(<span class="at">object =</span> pbmc_signac,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">group.by =</span>  <span class="st">"scDblFinder.class"</span>,</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(<span class="at">plotlist =</span> <span class="fu">list</span>(</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>  p_signac_filter<span class="sc">+</span><span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Total number of</span><span class="sc">\n</span><span class="st">fragments in peak"</span>),</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>  p_cluster<span class="sc">+</span><span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Signac clustering"</span>) ,</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>  p_cluster_scDblFinder<span class="sc">+</span><span class="fu">labs</span>(<span class="at">title=</span><span class="st">"scDblFinder"</span>),</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>  p_cluster_amulet<span class="sc">+</span><span class="fu">labs</span>(<span class="at">title=</span><span class="st">"amulet"</span>)</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"D"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-clustering_visual" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-clustering_visual-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="scATAC_seq_files/figure-html/fig-clustering_visual-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" data-glightbox="description: .lightbox-desc-4"><img src="scATAC_seq_files/figure-html/fig-clustering_visual-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-clustering_visual-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Doublet cells are not clustered
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-clustering_visual" class="quarto-xref">Figure&nbsp;4</a> A illustrates the cutoff values (blue lines) used in the Signac vignette to filter out low-quality cells. Despite this, many doublets identified by amulet remain in the original clustering solution, as shown in Figure <a href="#fig-clustering_visual" class="quarto-xref">Figure&nbsp;4</a> B. However, Figures <a href="#fig-clustering_visual" class="quarto-xref">Figure&nbsp;4</a> C and D reveal that doublet cells identified by both scDblFinder and amulet are dispersed throughout the UMAP and do not appear to influence the clustering structure.</p>
<p>While it is challenging to determine whether the presence of doublets significantly impacts downstream analyses without testing on additional datasets, doublet detection methods provide users with greater confidence in quality control, offering a robust approach to identify and account for potential artifacts..</p>
</section>
<section id="references" class="level3" data-number="2.8.2">
<h3 data-number="2.8.2" class="anchored" data-anchor-id="references"><span class="header-section-number">2.8.2</span> References</h3>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-R-GenomicRanges" class="csl-entry" role="listitem">
Aboyoun, Patrick, Hervé Pagès, and Michael Lawrence. 2024. <em>GenomicRanges: Representation and Manipulation of Genomic Intervals</em>. <a href="https://doi.org/10.18129/B9.bioc.GenomicRanges">https://doi.org/10.18129/B9.bioc.GenomicRanges</a>.
</div>
<div id="ref-R-MatrixGenerics" class="csl-entry" role="listitem">
Ahlmann-Eltze, Constantin, Peter Hickey, and Hervé Pagès. 2024. <em>MatrixGenerics: S4 Generic Summary Statistic Functions That Operate on Matrix-Like Objects</em>. <a href="https://doi.org/10.18129/B9.bioc.MatrixGenerics">https://doi.org/10.18129/B9.bioc.MatrixGenerics</a>.
</div>
<div id="ref-SingleCellExperiment2020" class="csl-entry" role="listitem">
Amezquita, Robert, Aaron Lun, Etienne Becht, Vince Carey, Lindsay Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span> <em>Nature Methods</em> 17: 137–45. <a href="https://www.nature.com/articles/s41592-019-0654-x">https://www.nature.com/articles/s41592-019-0654-x</a>.
</div>
<div id="ref-R-GenomeInfoDb" class="csl-entry" role="listitem">
Arora, Sonali, Martin Morgan, Marc Carlson, and Hervé Pagès. 2024. <em>GenomeInfoDb: Utilities for Manipulating Chromosome Names, Including Modifying Them to Follow a Particular Naming Style</em>. <a href="https://doi.org/10.18129/B9.bioc.GenomeInfoDb">https://doi.org/10.18129/B9.bioc.GenomeInfoDb</a>.
</div>
<div id="ref-R-matrixStats" class="csl-entry" role="listitem">
Bengtsson, Henrik. 2024. <em>matrixStats: Functions That Apply to Rows and Columns of Matrices (and to Vectors)</em>. <a href="https://github.com/HenrikBengtsson/matrixStats">https://github.com/HenrikBengtsson/matrixStats</a>.
</div>
<div id="ref-sp2013" class="csl-entry" role="listitem">
Bivand, Roger S., Edzer Pebesma, and Virgilio Gomez-Rubio. 2013. <em>Applied Spatial Data Analysis with <span>R</span>, Second Edition</em>. Springer, NY. <a href="https://asdar-book.org/">https://asdar-book.org/</a>.
</div>
<div id="ref-Seurat2018" class="csl-entry" role="listitem">
Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. <span>“Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.”</span> <em>Nature Biotechnology</em> 36: 411–20. <a href="https://doi.org/10.1038/nbt.4096">https://doi.org/10.1038/nbt.4096</a>.
</div>
<div id="ref-R-Biobase" class="csl-entry" role="listitem">
Gentleman, R., V. Carey, M. Morgan, and S. Falcon. 2024. <em>Biobase: Base Functions for Bioconductor</em>. <a href="https://doi.org/10.18129/B9.bioc.Biobase">https://doi.org/10.18129/B9.bioc.Biobase</a>.
</div>
<div id="ref-R-scDblFinder" class="csl-entry" role="listitem">
Germain, Pierre-Luc. 2024. <em>scDblFinder: scDblFinder</em>. <a href="https://doi.org/10.18129/B9.bioc.scDblFinder">https://doi.org/10.18129/B9.bioc.scDblFinder</a>.
</div>
<div id="ref-scDblFinder2022" class="csl-entry" role="listitem">
Germain, Pierre-Luc, Aaron Lun, Carlos Garcia Meixide, Will Macnair, and Mark D. Robinson. 2022. <span>“Doublet Identification in Single-Cell Sequencing Data Using scDblFinder.”</span> <em>F1000research</em>. <a href="https://doi.org/10.12688/f1000research.73600.2">https://doi.org/10.12688/f1000research.73600.2</a>.
</div>
<div id="ref-Seurat2021" class="csl-entry" role="listitem">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. <span>“Integrated Analysis of Multimodal Single-Cell Data.”</span> <em>Cell</em>. <a href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div id="ref-Seurat2023" class="csl-entry" role="listitem">
Hao, Yuhan, Tim Stuart, Madeline H Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, et al. 2023. <span>“Dictionary Learning for Integrative, Multimodal and Scalable Single-Cell Analysis.”</span> <em>Nature Biotechnology</em>. <a href="https://doi.org/10.1038/s41587-023-01767-y">https://doi.org/10.1038/s41587-023-01767-y</a>.
</div>
<div id="ref-heumos2023best" class="csl-entry" role="listitem">
Heumos, Lukas, Anna C Schaar, Christopher Lance, Anastasia Litinetskaya, Felix Drost, Luke Zappia, Malte D Lücken, et al. 2023. <span>“Best Practices for Single-Cell Analysis Across Modalities.”</span> <em>Nature Reviews Genetics</em> 24 (8): 550–72.
</div>
<div id="ref-R-SeuratObject" class="csl-entry" role="listitem">
Hoffman, Paul, Rahul Satija, David Collins, Yuhan Hao, Austin Hartman, Gesmira Molla, Andrew Butler, and Tim Stuart. 2024. <em>SeuratObject: Data Structures for Single Cell Data</em>. <a href="https://satijalab.github.io/seurat-object/">https://satijalab.github.io/seurat-object/</a>.
</div>
<div id="ref-Biobase2015" class="csl-entry" role="listitem">
Huber, W., V. J. Carey, R. Gentleman, S. Anders, M. Carlson, B. S. Carvalho, H. C. Bravo, et al. 2015a. <span>“<span>O</span>rchestrating High-Throughput Genomic Analysis with <span>B</span>ioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html">http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html</a>.
</div>
<div id="ref-BiocGenerics2015" class="csl-entry" role="listitem">
Huber, W., Carey, V. J., Gentleman, R., Anders, et al. 2015b. <span>“<span>O</span>rchestrating High-Throughput Genomic Analysis with <span>B</span>ioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html">http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html</a>.
</div>
<div id="ref-R-ggpubr" class="csl-entry" role="listitem">
Kassambara, Alboukadel. 2023. <em>Ggpubr: Ggplot2 Based Publication Ready Plots</em>. <a href="https://rpkgs.datanovia.com/ggpubr/">https://rpkgs.datanovia.com/ggpubr/</a>.
</div>
<div id="ref-GenomicRanges2013" class="csl-entry" role="listitem">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin Morgan, and Vincent Carey. 2013a. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em><span>PLoS</span> Computational Biology</em> 9. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-IRanges2013" class="csl-entry" role="listitem">
———. 2013b. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em><span>PLoS</span> Computational Biology</em> 9. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-R-SingleCellExperiment" class="csl-entry" role="listitem">
Lun, Aaron, and Davide Risso. 2024. <em>SingleCellExperiment: S4 Classes for Single Cell Data</em>.
</div>
<div id="ref-R-SummarizedExperiment" class="csl-entry" role="listitem">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2024. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://doi.org/10.18129/B9.bioc.SummarizedExperiment">https://doi.org/10.18129/B9.bioc.SummarizedExperiment</a>.
</div>
<div id="ref-R-IRanges" class="csl-entry" role="listitem">
Pagès, Hervé, Patrick Aboyoun, and Michael Lawrence. 2024. <em>IRanges: Foundation of Integer Range Manipulation in Bioconductor</em>. <a href="https://doi.org/10.18129/B9.bioc.IRanges">https://doi.org/10.18129/B9.bioc.IRanges</a>.
</div>
<div id="ref-R-S4Vectors" class="csl-entry" role="listitem">
Pagès, Hervé, Michael Lawrence, and Patrick Aboyoun. 2024. <em>S4Vectors: Foundation of Vector-Like and List-Like Containers in Bioconductor</em>. <a href="https://doi.org/10.18129/B9.bioc.S4Vectors">https://doi.org/10.18129/B9.bioc.S4Vectors</a>.
</div>
<div id="ref-sp2005" class="csl-entry" role="listitem">
Pebesma, Edzer J., and Roger Bivand. 2005. <span>“Classes and Methods for Spatial Data in <span>R</span>.”</span> <em>R News</em> 5 (2): 9–13. <a href="https://CRAN.R-project.org/doc/Rnews/">https://CRAN.R-project.org/doc/Rnews/</a>.
</div>
<div id="ref-R-sp" class="csl-entry" role="listitem">
Pebesma, Edzer, and Roger Bivand. 2024. <em>Sp: Classes and Methods for Spatial Data</em>. <a href="https://github.com/edzer/sp/">https://github.com/edzer/sp/</a>.
</div>
<div id="ref-R-base" class="csl-entry" role="listitem">
R Core Team. 2024. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-R-Seurat" class="csl-entry" role="listitem">
Satija, Rahul. 2024. <em>Seurat: Tools for Single Cell Genomics</em>. <a href="https://satijalab.org/seurat">https://satijalab.org/seurat</a>.
</div>
<div id="ref-Seurat2015" class="csl-entry" role="listitem">
Satija, Rahul, Jeffrey A Farrell, David Gennert, Alexander F Schier, and Aviv Regev. 2015. <span>“Spatial Reconstruction of Single-Cell Gene Expression Data.”</span> <em>Nature Biotechnology</em> 33: 495–502. <a href="https://doi.org/10.1038/nbt.3192">https://doi.org/10.1038/nbt.3192</a>.
</div>
<div id="ref-Seurat2019" class="csl-entry" role="listitem">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive Integration of Single-Cell Data.”</span> <em>Cell</em> 177: 1888–1902. <a href="https://doi.org/10.1016/j.cell.2019.05.031">https://doi.org/10.1016/j.cell.2019.05.031</a>.
</div>
<div id="ref-R-Signac" class="csl-entry" role="listitem">
Stuart, Tim, and Avi Srivastava. 2024. <em>Signac: Analysis of Single-Cell Chromatin Data</em>. <a href="https://github.com/stuart-lab/signac">https://github.com/stuart-lab/signac</a>.
</div>
<div id="ref-Signac2021" class="csl-entry" role="listitem">
Stuart, Tim, Avi Srivastava, Shaista Madad, Caleb Lareau, and Rahul Satija. 2021. <span>“Single-Cell Chromatin State Analysis with Signac.”</span> <em>Nature Methods</em>. <a href="https://doi.org/10.1038/s41592-021-01282-5">https://doi.org/10.1038/s41592-021-01282-5</a>.
</div>
<div id="ref-R-BiocGenerics" class="csl-entry" role="listitem">
Team, The Bioconductor Dev. 2024. <em>BiocGenerics: S4 Generic Functions Used in Bioconductor</em>. <a href="https://doi.org/10.18129/B9.bioc.BiocGenerics">https://doi.org/10.18129/B9.bioc.BiocGenerics</a>.
</div>
<div id="ref-thibodeau2021amulet" class="csl-entry" role="listitem">
Thibodeau, Asa, Alper Eroglu, Christopher S McGinnis, Nathan Lawlor, Djamel Nehar-Belaid, Romy Kursawe, Radu Marches, et al. 2021. <span>“AMULET: A Novel Read Count-Based Method for Effective Multiplet Detection from Single Nucleus ATAC-Seq Data.”</span> <em>Genome Biology</em> 22: 1–19.
</div>
<div id="ref-ggplot22016" class="csl-entry" role="listitem">
Wickham, Hadley. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. Springer-Verlag New York. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-R-ggplot2" class="csl-entry" role="listitem">
Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen, Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, Dewey Dunnington, and Teun van den Brand. 2024. <em>Ggplot2: Create Elegant Data Visualisations Using the Grammar of Graphics</em>. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-R-dplyr" class="csl-entry" role="listitem">
Wickham, Hadley, Romain François, Lionel Henry, Kirill Müller, and Davis Vaughan. 2023. <em>Dplyr: A Grammar of Data Manipulation</em>. <a href="https://dplyr.tidyverse.org">https://dplyr.tidyverse.org</a>.
</div>
<div id="ref-R-ggvenn" class="csl-entry" role="listitem">
Yan, Linlin. 2023. <em>Ggvenn: Draw Venn Diagram by Ggplot2</em>.
</div>
</div>
<details>
<summary>
session Info
</summary>
</details>


<div class="hidden" aria-hidden="true">
<span class="glightbox-desc lightbox-desc-1">Figure&nbsp;1: Relationship between Peaks and TSS enrichment.</span>
<span class="glightbox-desc lightbox-desc-2">Figure&nbsp;2: </span>
<span class="glightbox-desc lightbox-desc-3">Figure&nbsp;3: Amulet more correlats with ‘Total number of fragments in peaks’</span>
<span class="glightbox-desc lightbox-desc-4">Figure&nbsp;4: Doublet cells are not clustered</span>
</div>
</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = true;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Copyright 2025, Rishi Das Roy</p>
</div>   
    <div class="nav-footer-center">

<div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
window.onload = () => {
  lightboxQuarto.on('slide_before_load', (data) => {
    const { slideIndex, slideNode, slideConfig, player, trigger } = data;
    const href = trigger.getAttribute('href');
    if (href !== null) {
      const imgEl = window.document.querySelector(`a[href="${href}"] img`);
      if (imgEl !== null) {
        const srcAttr = imgEl.getAttribute("src");
        if (srcAttr && srcAttr.startsWith("data:")) {
          slideConfig.href = srcAttr;
        }
      }
    } 
  });

  lightboxQuarto.on('slide_after_load', (data) => {
    const { slideIndex, slideNode, slideConfig, player, trigger } = data;
    if (window.Quarto?.typesetMath) {
      window.Quarto.typesetMath(slideNode);
    }
  });

};
          </script>




</body></html>